jfr/src/main/java/io/netty/jfr/JfrChannelHandler.java

```diff
@@ -0,0 +1,417 @@
+        writeEvent.begin();
+        writeEvent.setMsg(msg);
+        ctx.write(msg, promise);
+        writeEvent.end();
+        if (writeEvent.shouldCommit()) {
+            writeEvent.setChannel(ctx.channel());
+            writeEvent.commit();
+        }
+    }
+
+    @Override
+    public void channelWritabilityChanged(ChannelHandlerContext ctx) {
+        final ChannelWritabilityChangedEvent channelWritabilityChangedEvent = new ChannelWritabilityChangedEvent();
+        channelWritabilityChangedEvent.begin();
+        channelWritabilityChangedEvent.setWasWritable(ctx.channel().isWritable());
+        ctx.fireChannelWritabilityChanged();
+        channelWritabilityChangedEvent.end();
+        if (channelWritabilityChangedEvent.shouldCommit()) {
+            channelWritabilityChangedEvent.setChannel(ctx.channel());
+            channelWritabilityChangedEvent.setIsWritable(ctx.channel().isWritable());
+            channelWritabilityChangedEvent.commit();
+        }
+    }
+
+    @Override
+    public void flush(ChannelHandlerContext ctx) {
+        final FlushEvent flushEvent = new FlushEvent();
+        flushEvent.begin();
+        ctx.flush();
+        flushEvent.end();
+        if (flushEvent.shouldCommit()) {
+            flushEvent.setChannel(ctx.channel());
+            flushEvent.commit();
+        }
+    }
+
+    // Event types
+
+    @Category({ "Netty", "Channel" })
+    @Enabled(true)
+    @Threshold("1 us")
+    private abstract static class ChannelEvent extends Event {
+        @Label("Channel")
+        String channel;
+        @Label("Long Channel ID")
+        String longChannelId;
+        @Label("Short Channel ID")
+        String shortChannelId;
+
+        void setChannel(Channel channel) {
+            this.channel = String.valueOf(channel);
+            //TODO: More specific channel fields?
+            this.longChannelId = channel.id().asLongText();
+            this.shortChannelId = channel.id().asShortText();
+        }
+    }
+
+    @Threshold("20 ms")
+    private abstract static class ChannelIOEvent extends ChannelEvent {
+    }
+
+    @Label("Bind")
+    @Name("io.netty.channel.Bind")
+    private static final class BindEvent extends ChannelEvent {
+        @Label("Local Address")
+        String localAddress;
+
+        void setLocalAddress(SocketAddress localAddress) {
+            this.localAddress = String.valueOf(localAddress);
+        }
+    }
+
+    @Label("Write")
+    @Name("io.netty.channel.Write")
+    private static final class WriteEvent extends ChannelIOEvent {
+        @Label("Message")
+        String msg;
+
+        void setMsg(Object msg) {
+            this.msg = String.valueOf(msg);
+        }
+    }
+
+    @Label("Channel Read")
+    @Name("io.netty.channel.ChannelRead")
+    private static final class ChannelReadEvent extends ChannelIOEvent {
+        @Label("Message")
+        String msg;
+
+        void setMsg(Object msg) {
+            this.msg = String.valueOf(msg);
```

Comment:
Could storing the read or write message lead to really big events, and therefore really big JFR recordings ?

Maybe instead the event could store the size of the message, yet this looks cumbersome as the message is `Object` typed.

Reply:
@normanmaurer Any ideas for getting something useful but not too big from msg?